### PR TITLE
refactor: ownership redesign

### LINKED ISSUE #108: Remove "transferOwnership" functionality
## Description

As discussed in https://github.com/PaulRBerg/prb-proxy/discussions/80, the rationale for keeping `transferOwnership` was to make it an escape hatch for wallet migration purposes.

However, during their recent audit, Cantina discovered many undesirable second-order effects related to this ownership transfer functionality - so many that I have decided to remove it altogether.

Not letting ownership be transferred is not so bad, considering that the `execute` function allows the owner to perform any arbitrary computation. Migrating the wallet is still possible - with a few hops.

### Cantina

- https://github.com/cantinasec/review-sablier2/issues/2
- https://github.com/cantinasec/review-sablier2/issues/7
- https://github.com/cantinasec/review-sablier2/issues/12
- https://github.com/cantinasec/review-sablier2/issues/14
- https://github.com/cantinasec/review-sablier2/issues/22
- https://github.com/cantinasec/review-sablier2/issues/24

### Omniscia

- [PRR-01M: Inexistent Verification of Acceptance](https://omniscia.io/reports/sablier-prb-proxy-647f15164f25e70014286487/manual-review/PRBProxyRegistry-PRR#PRR-01M)
- [PRR-03M: Insecure Post-Execution Transfer](https://omniscia.io/reports/sablier-prb-proxy-647f15164f25e70014286487/manual-review/PRBProxyRegistry-PRR#PRR-03M)

### PR DESCRIPTION
**Pull request by PaulRBerg**
Closes #108, addressing all of the audit report findings mentioned there.

### CODE REVIEW

File: src/PRBProxy.sol
Comment (andreivladbrg) [inline src/PRBProxy.sol:109]: Aren't the braces used when {contract.foo}?

**Review by andreivladbrg — APPROVED** [review]
I can say that the design is cleaner now.

Not sure if it's a good idea(it would slightly increase the gas) but what do you think about changing the internal `_deploy` function to: `function _deploy(ConstructorParams calldata params)`. This way you will be able to remove duplicated code from `_deploy` and `deployAndExecute`.

File: src/PRBProxy.sol
Comment (PaulRBerg) [reply src/PRBProxy.sol:109]: Good catch - I will change this to:

````
`execute`
````

**Comment by PaulRBerg** [comment]
Thanks for the review. I agree that the new design is clearer - and safer.

Interesting idea to pass `ConstructorParams` to `_deploy,` but I suggest we don't do it because it would increase the gas cost by a non-trivial amount (not just "slightly"). ABI-encoding and decoding structs is an expensive operation. And gas costs are very important for deployments (on Ethereum Mainnet), so I would prefer to save any little that we can.

Both functions are well-tested, so we have a pretty good idea that they work as expected.

**Comment by andreivladbrg** [comment]
Just noticed

https://github.com/PaulRBerg/prb-proxy/blob/17f65af6ecdad19d3807929785ab7bf255d8f7f4/src/PRBProxyRegistry.sol#L132-L134

Given that the only variable used here is the owner, this can be optimized to:

```solidity
```

**Comment by PaulRBerg** [comment]
Interesting optimization idea @andreivladbrg but the two implementations cost the same gas when only the `owner` is set. See my conversation with ChatGPT:

https://chat.openai.com/share/3187fa95-2732-4fe8-95c0-539e259fc835

Also, assigning only the `owner` to zero is less secure than deleting the entire struct because of the possibility of reentrancy.

**Review by tibbarytsur — COMMENTED** [review]
Set the proxies[owner] to a temp value before creating the proxy to avoid reentrancy for the same owner
https://github.com/PaulRBerg/prb-proxy/blob/17f65af6ecdad19d3807929785ab7bf255d8f7f4/src/PRBProxyRegistry.sol#L95-L99
`noProxy(owner)` modifier would pass otherwise when calling `deployFor` on the registry with the same  owner creating 2 proxies with the same owner set in the proxy itself, and the first one set in the registry.

**Comment by tibbarytsur** [comment]
All constructor params should be set here (target and data to 0), not just the owner 
https://github.com/PaulRBerg/prb-proxy/blob/17f65af6ecdad19d3807929785ab7bf255d8f7f4/src/PRBProxyRegistry.sol#L132

`deployAndExecute` can reenter in `deployFor` (via a target contract) which would only change the `constructorParams.owner` leaving `target` and `data` intact. The PRBProxy constructor would then get the constructor params from the registry including the old `target` and `data` and execute it on the second proxy (this time doing something else like setting token approvals).

**Comment by andreivladbrg** [comment]
> See my conversation with ChatGPT

I don't think that using GPT for this specific use case is the best choice, tested this in remix it was actually a difference by few thousands.

> the possibility of reentrancy.

Right, it is safer this way.

**Comment by PaulRBerg** [comment]
Thanks for your review, both!

> `noProxy(owner) `modifier would pass otherwise when calling `deployFor` on the registry with the same owner creating 2 proxies with the same owner set in the proxy itself
> ...
> `deployAndExecute` can reenter in `deployFor` (via a target contract) which would only change the `constructorParams.owner` leaving `target` and `data` intact. 

As discussed on Discord, this cannot happen under the current design - happy to take a call tomorrow to walk through the low-level steps.

> All constructor params should be set here (target and data to 0), not just the owner

Agree with this proposal - it would be a good precautionary step even if the attack above cannot happen.

> I don't think that using GPT for this specific use case is the best choice, tested this in remix and it was actually a difference by few thousands.

You may be right about ChatGPT (the 2021 cut-off), but did you use [via IR](https://github.com/ethereum/remix-project/issues/2350#issuecomment-1365919609) and the same number of optimizer runs in your test?

> Right, it is safer to use `delete`.

Yep

**Comment by andreivladbrg** [comment]
> but did you use https://github.com/ethereum/remix-project/issues/2350#issuecomment-1365919609 and the same number of optimizer runs in your test

Yes

**Comment by PaulRBerg** [comment]
@tibbarytsur and @andreivladbrg have reviewed this PR, but I will also wait for Zach's confirmation before merging this in.

**Pull request merged**
